src/Microsoft.AspNet.Razor/Tokenizer/CSharpKeywordDetector.cs

```diff
@@ -88,7 +88,8 @@ internal static class CSharpKeywordDetector
             { "interface", CSharpKeyword.Interface },
             { "break", CSharpKeyword.Break },
             { "checked", CSharpKeyword.Checked },
-            { "namespace", CSharpKeyword.Namespace }
+            { "namespace", CSharpKeyword.Namespace },
+            { "when", CSharpKeyword.When }
```

Comment:
This makes `when` a keyword in a few irrelevant contexts. Suggest instead looking for an identifier with the name `"when"` in `FilterableCatchBlock()`. (General problem that many supposed keywords in this collection are actually contextual. Let's not make that worse.)